### PR TITLE
otelrestful: support WithPublicEndpoint and WithPublicEndpointFn

### PR DESCRIPTION
**Pull request by frzifus**
The implementation is strongly inspired by otlphttp.

Closes https://github.com/open-telemetry/opentelemetry-go-contrib/issues/3562

### CODE REVIEW

**Comment by dmathieu** [comment]
otelhttp also has `WithPublicEndpointFn`. How about supporting both?

**Comment by frzifus** [comment]
sure - shouldn't be to much effort.

**Comment by dmathieu** [comment]
Thank you! ❤️ 
Could you add tests for the new method?

**Comment by frzifus** [comment]
done @dmathieu. Could have please have a look again? :)

**Review by dmathieu — APPROVED** [review]

**Review by pellared — APPROVED** [review]

**Comment by frzifus** [comment]
Is there something missing? :)

File: instrumentation/github.com/emicklei/go-restful/otelrestful/config.go [hunk 1]
```diff
@@ -56,3 +69,14 @@ func WithTracerProvider(provider oteltrace.TracerProvider) Option {
 		}
 	})
 }
+
+// WithPublicEndpointFn runs with every request, and allows conditionnally
+// configuring the Handler to link the span with an incoming span context. If
+// this option is not provided or returns false, then the association is a
+// child association instead of a link.
+// Note: WithPublicEndpoint takes precedence over WithPublicEndpointFn.
```
Comment (MrAlias) [inline instrumentation/github.com/emicklei/go-restful/otelrestful/config.go:77]:
```suggestion
// Note: [WithPublicEndpoint] takes precedence over WithPublicEndpointFn.
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -11,6 +11,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 ### Added
 
 - Add the new `go.opentelemetry.io/contrib/instrgen` package to provide auto-generated source code instrumentation. (#3068)
+- Support `WithPublicEndpoint` and `WithPublicEndpointFn` in otelrestfulCont. (#3563)
```
Comment (MrAlias) [inline CHANGELOG.md:14]:
```suggestion
- The `WithPublicEndpoint` and `WithPublicEndpointFn` options in `go.opentelemetry.io/contrib/instrumentation/github.com/emicklei/go-restful/otelrestful`. (#3563)
```

**Review by MrAlias — APPROVED** [review]

**Pull request merged**